app/src/main/java/com/globant/activities/CharactersActivity.kt

```diff
@@ -13,22 +13,60 @@ import com.globant.utils.Data
 import com.globant.utils.Event
 import com.globant.utils.Status
 import com.globant.viewmodels.CharactersViewModel
+import kotlinx.android.synthetic.main.activity_characters.button_clear
+import kotlinx.android.synthetic.main.activity_characters.button_local_data
+import kotlinx.android.synthetic.main.activity_characters.button_refresh
 import kotlinx.android.synthetic.main.activity_characters.progress_bar
 import kotlinx.android.synthetic.main.activity_characters.recycler_view_characters
 import org.koin.androidx.viewmodel.ext.android.viewModel
 
 class CharactersActivity : AppCompatActivity() {
 
     private val viewModel by viewModel<CharactersViewModel>()
-    private lateinit var adapter: CharacterAdapter
+    private val adapter = CharacterAdapter { character ->
+        val characterFragment = CharacterDetailFragment.newInstance(character.id)
+        characterFragment.show(supportFragmentManager, getString(R.string.tag))
+    }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_characters)
-        viewModel.getAllCharacters()
+        button_refresh.setOnClickListener {
+            viewModel.getAllCharacters()
+        }
+        button_local_data.setOnClickListener {
+            viewModel.getLocalCharacters()
+        }
+        button_clear.setOnClickListener {
+            adapter.update(emptyList())
+        }
+        viewModel.localDataState.observe(::getLifecycle, ::showLocalData)
         viewModel.mainState.observe(::getLifecycle, ::updateUI)
     }
 
+    private fun showLocalData(charactersData: Event<Data<List<MarvelCharacter>>>) {
+        when (charactersData.peekContent().responseType) {
+            Status.GetLocalCharacterError -> {
+                progress_bar.visibility = View.GONE
```

Comment:
to grant more legibility when, in the possibilities, you have more than 2 lines you can create a private method with those lines

Reply:
Done